tools/templatetest/main.go

```diff
@@ -0,0 +1,318 @@
+{{.NestingTest.A1.B1.C1}} {{.NestingTest.A1.B2.C2}} {{.NestingTest.A1.B2.C3}}
+{{- range .RangeNest1}}
+{{- range .RangeNest2}}
+{{.LGTM}}
+{{- end}}
+{{- end}}
+{{- end}}
+`
+
+type TemplateCallTree struct {
+	listType   *TemplateCallTree
+	structType *TemplateCallTreeEntry
+}
+
+func NewTemplateCallTree() *TemplateCallTree {
+	return &TemplateCallTree{
+		listType:   nil,
+		structType: nil,
+	}
+}
+
+func (c *TemplateCallTree) String() string {
+	out := ""
+
+	if c.listType != nil {
+		out += c.listType.structType.Fprint(0)
+	}
+
+	if c.structType != nil {
+		out += c.structType.Fprint(0)
+	}
+
+	return out
+}
+
+func (c *TemplateCallTree) List() *TemplateCallTree {
+	if c.listType == nil {
+		c.listType = NewTemplateCallTree()
+	}
+
+	return c.listType
+}
+
+func (c *TemplateCallTree) IsValid() bool {
+	switch {
+	case c.listType == nil && c.structType == nil:
+		return true
+	case c.listType != nil && c.structType == nil:
+		return c.listType.IsValid()
+	case c.listType == nil && c.structType != nil:
+		return c.structType.IsValid()
+	default:
+		return false
+	}
+}
+
+func (c *TemplateCallTree) Struct() *TemplateCallTreeEntry {
+	if c.structType == nil {
+		c.structType = NewTemplateCallTreeEntry()
+	}
+
+	return c.structType
+}
+
+type TemplateCallTreeEntry struct {
+	fields map[string]*TemplateCallTree
+}
+
+func NewTemplateCallTreeEntry() *TemplateCallTreeEntry {
+	return &TemplateCallTreeEntry{
+		fields: make(map[string]*TemplateCallTree),
+	}
+}
+
+func (c *TemplateCallTreeEntry) IsValid() bool {
+	for _, t := range c.fields {
+		if !t.IsValid() {
+			return false
+		}
+	}
+
+	return true
+}
+
+func (c *TemplateCallTreeEntry) Fprint(depth int) string {
+	out := ""
+
+	for key, value := range c.fields {
+		if value.listType != nil {
+			out += ident(fmt.Sprintf("- []%v:\n", key), depth)
+			out += value.listType.structType.Fprint(depth + 1)
+		}
+		if value.structType != nil {
+			out += ident(fmt.Sprintf("- %v:\n", key), depth)
+			out += value.structType.Fprint(depth + 1)
+		}
+		if value.listType == nil && value.structType == nil {
+			out += ident(fmt.Sprintf("- %v\n", key), depth)
+		}
+	}
+
+	return out
+}
+
+const spacesPerDepth = 4
+
+func ident(value string, depth int) string {
+	return fmt.Sprintf("%*s%s", depth*spacesPerDepth, "", value)
+}
+
+func main() {
+	if err := inner(); err != nil {
+		fmt.Println(err)
+	}
+}
+
+func inner() error {
+	parseTree, err := parse.Parse("templ", describeTemplate, "{{", "}}", templateFuncs)
+	if err != nil {
+		return err
+	}
+
+	templ := parseTree["templ"]
+	root := NewTemplateCallTree()
+
+	if err := buildTree(root, templ.Root); err != nil {
+		return err
+	}
+
+	fmt.Println(root)
+
+	return nil
+}
+
+//nolint:gocyclo
+func buildTree(root *TemplateCallTree, node parse.Node) error {
+	if IsNil(node) {
+		return nil
+	}
+
+	buildAndMergeTrees := func(nodes ...parse.Node) error {
+		for _, node := range nodes {
+			if err := buildTree(root, node); err != nil {
+				return err
+			}
+		}
+
```

Comment:
[nit] some extra lines are not needed across the code, if we could remove, would be nice